### PR TITLE
Find with options

### PR DESCRIPTION
**Pull request by SimonRichardson**
Update the find command to take into consideration the filter logic.
Unfortunately, there are some filtering logic that is present in the api
doc that isn't currently supported, although they will be added and
some are only available on staging API.

The following will be added, but are still being implemented

 - platforms
 - relation-requires
 - relation-provides

The following is in staging

 - publisher

I'm going to leave the ones to be implemented on the server, but hide
them in the cmd, so we can just add them when they become available.

## QA steps

```sh
$ juju bootstrap lxd test
$ juju find --type=charm juju
$ juju find --type=bundle kube
$ juju find --channel=stable juju
$ juju find juju
```

### CODE REVIEW

**Comment by hmlanigan** [comment]
QA: 

$ juju find --type bundle
No search term specified. Here are some interesting charms:     <-- Not relevant with the different new flags.

------------

$ juju find --publisher bootstack-charmers
ERROR unexpected api error attempting to query charm or bundle from the charmhub store

{"error-list":[{"code":"api-error","message":"Bad parameters ['publisher']"}]}

------------

These two may be upstream in charmhub?

$ juju find --channel 2.0/edge
No matching charms for ""       <-- the juju-qa-test charm in production has a 2.0/edge open channel.

$ juju find --category kubeflow
No matching charms for ""       <--- the admission-webhook charm has a tag of kubeflow

-------------

File: cmd/juju/charmhub/find.go [hunk 1]
```diff
@@ -70,13 +79,19 @@ func (c *findCommand) Info() *cmd.Info {
...
-	c.charmHubCommand.SetFlags(f)
+	c.ModelCommandBase.SetFlags(f)
 
 	c.out.AddFlags(f, "tabular", map[string]cmd.Formatter{
 		"yaml":    cmd.FormatYaml,
 		"json":    cmd.FormatJson,
 		"tabular": c.formatter,
 	})
+
+	f.StringVar(&c.category, "category", "", `filter by a category name`)
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply cmd/juju/charmhub/find.go:90]: They use category for find... :tada:

**Comment by SimonRichardson** [comment]
Replied inline:

> $ juju find --type bundle
> No search term specified. Here are some interesting charms: <-- Not relevant with the different new flags.

I'll fix this.

> $ juju find --publisher bootstack-charmers
> ERROR unexpected api error attempting to query charm or bundle from the charmhub store

> {"error-list":[{"code":"api-error","message":"Bad parameters ['publisher']"}]}

Still waiting for this to land correctly on staging.

> These two may be upstream in charmhub?

> $ juju find --channel 2.0/edge
> No matching charms for "" <-- the juju-qa-test charm in production has a 2.0/edge open channel.

> $ juju find --category kubeflow
> No matching charms for "" <--- the admission-webhook charm has a tag of kubeflow

I'll remove `from ""` the error message.

> Individually these 2 flags produce results, but no results when combined. All of the results from juju find --channel=beta juju are charms.

> $ juju find --channel=beta --category charm
> No matching charms for ""

This isn't our issue, but I should ask about this.

**Review by manadart — APPROVED** [review]
Nice job.

**Comment by SimonRichardson** [comment]
!!build!!

**Comment by SimonRichardson** [comment]
$$merge$$

**Comment by SimonRichardson** [comment]
$$merge$$

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**